Libraries/LLM/Lora.swift

```diff
@@ -0,0 +1,564 @@
+///
+/// This layer implements the LoRA capabilities, specifically:
+///
+/// - converting `Linear` or `QuantizedLinear` layers to ``LoRALinear``
+/// - converting ``LoRALinear`` back to `Linear` or `QuantizedLinear`
+/// - implementing the LoRA evaluation
+///
+/// This is not typically used directly -- ``LoRATrain/convert(model:layers:)`` is used to
+/// add the adapter layers to a given model.
+///
+/// ### See Also
+/// - [LoRA: Low-Rank Adaptation of Large Language Models](https://arxiv.org/abs/2106.09685)
+/// - [QLoRA: Efficient Finetuning of Quantized LLMs](https://arxiv.org/abs/2305.14314)
+/// - ``LoRALayer``
+/// - ``LoRATrain/convert(model:layers:)``
+/// - ``LoRATrain/fuse(model:deQuantize:)``
+public class LoRALinear: Linear, LoRAConvertToLinear {
+
+    let scale: Float
+
+    @ParameterInfo(key: "lora_a") var loraA: MLXArray
+    @ParameterInfo(key: "lora_b") var loraB: MLXArray
+
+    required public init(
+        _ inputDimensions: Int, _ outputDimensions: Int, rank: Int = 8, bias: Bool = false,
+        scale: Float = 20.0, linear: Linear
+    ) {
+        // Scale for low-rank update
+        self.scale = scale
+
+        // Low rank lora weights
+        let loraScale = 1 / sqrt(Float(inputDimensions))
+        self._loraA.wrappedValue = MLXRandom.uniform(
+            low: -loraScale, high: loraScale, [inputDimensions, rank])
+        self._loraB.wrappedValue = MLXArray.zeros([rank, outputDimensions])
+        
+        super.init(weight: linear.weight, bias: linear.bias)
+    }
+
+    /// Convert a `Linear` or `QuantizedLinear` layer into ``LoRALinear``.
+    ///
+    /// This is typically called via ``LoRATrain/convert(model:layers:)``.
+    ///
+    /// ### See Also
+    /// - ``LoRATrain/convert(model:layers:)``
+    public static func from(linear: Linear, rank: Int = 8) -> Linear {
+        if let linear = linear as? QuantizedLinear {
+            return QLoRALinear.from(linear: linear, rank: rank)
+        }
+        let (outputDimensions, inputDimensions) = linear.shape
+        return LoRALinear(inputDimensions, outputDimensions, rank: rank, linear: linear)
+    }
+
+    /// Convert a ``LoRALinear`` back into a fused `Linear` or `QuantizedLinear` layer.
+    ///
+    /// This is typically called via ``LoRATrain/fuse(model:deQuantize:)``.
+    ///
+    /// ### See Also
+    /// - ``LoRATrain/fuse(model:deQuantize:)``
+    public func toLinear(deQuantize: Bool = false) -> Linear {
+        let dtype = weight.dtype
+        let loraB = (scale * loraB.T).asType(dtype)
+        let loraA = loraA.T.asType(dtype)
+        return Linear(weight: weight + matmul(loraB, loraA), bias: bias)
+    }
+
+    public override func callAsFunction(_ x: MLXArray) -> MLXArray {
+        let y = super.callAsFunction(x.asType(weight.dtype))
+        let z = matmul(matmul(x, self.loraA), self.loraB)
+        return y + scale * z
+    }
+}
+
+public class QLoRALinear: QuantizedLinear, LoRAConvertToLinear {
```

Comment:
@awni and here is the QLoRALinear for QuantizedLinear

I still need to update the docs, but now we don't need the special markers in the model code.  We are also not conditional on the type any more at the cost of a small amount of code duplication so I think it looks cleaner and easier to understand.

Reply:
Looks great to me!